openmls/src/group/core_group/validation.rs

```diff
@@ -171,6 +165,59 @@ impl CoreGroup {
                 return Err(ProposalValidationError::InsufficientCapabilities);
             }
 
+            // Leaf node validation.
+            {
+                let leaf_node = add_proposal.add_proposal().key_package().leaf_node();
+
+                let filtered_group_signature_keys = {
+                    let mut tmp = Vec::new();
+
+                    for Member {
+                        index,
+                        signature_key,
+                        ..
+                    } in self.treesync().full_leave_members()
+                    {
+                        // Exclude signature keys that will be removed.
+                        let has_remove_proposal = proposal_queue
+                            .remove_proposals()
+                            .any(|p| p.remove_proposal().removed == index);
+
+                        if !has_remove_proposal {
+                            tmp.push(SignaturePublicKey::from(signature_key));
+                        }
+                    }
+
+                    tmp
+                };
+
+                leaf_node
+                    .validate_in_key_package(
+                        self.version(),
+                        self.ciphersuite(),
+                        self.group_context_extensions().required_capabilities(),
+                        &filtered_group_signature_keys,
+                        &self.group_encryption_keys(),
+                        &self.members_supported_credentials(),
+                        &self.currently_in_use(),
+                    )
+                    .map_err(|error| match error {
```

Comment:
Shouldn't all this be handled by the `into`?

Reply:
In theorie yes. But there are differences in the errors that are returns regarding duplicate signature keys in a proposal list v.s. duplicate signature keys in a leaf node and I wanted to make sure that I don't break the tests when I can prevent it. We can hopefully remove all these things when things are settled and we unify error handling.